### ISSUE TITLE
readFileSync with 'utf-8' option broken by node v20.5.0

### ISSUE THREAD

**Issue by lg250137** [tschaub/mock-fs#377]
Node v20.5.0 introduced an optimization to `fs.readFileSync` for utf-8 files.  Unfortunately, this breaks mock-fs.  Here is a minimal test case that works with Node v20.4.0 but breaks with Node v20.5.0:

```
import mockFs from "mock-fs";
import { test } from "node:test";
import { readFileSync } from "fs";
import assert from "node:assert/strict";

test("mockFs", () => {
    mockFs({ "file": "contents" })
    const fileContents = readFileSync("file", 'utf-8').toString()
    assert.equal(fileContents, "contents")
})

```

In Node v20.5.0, I get the error ` Error: ENOENT: no such file or directory, open 'file'`.
The above test case also works in Node v20.5.0 if I remove the `'utf-8'` option from `readFileSync`.  This issue specifically occurs for `readFileSync` with `'utf-8'` option.

Here is the relevant commit: https://github.com/nodejs/node/pull/48658

**Comment by brendandburns** [comment]
fwiw, anyone who hits this, you can replace `fs.readFileSync(foo, 'utf-8')` with `fs.readFileSync(foo).toString('utf-8')`  and it will fix the mocking.

Of course you will lose the performance improvements in the nodejs change, so it's probably worth fixing mock-fs.

**Comment by marikaner** [comment]
I guess this PR relates to this as well: https://github.com/tschaub/mock-fs/pull/381/files.

**Comment by cihandeniz** [comment]
`readFileSync('...').toString('utf8')` was an ok workaround for me, but I had issues with `existsSync` and `writeFileSync` as well. I'm posting my temp solution in case anyone else has the same problem.

```javascript
// fsExt.js
import fs from "fs";

function existsSync(path) {
  try {
    fs.statSync(path);
    return true;
  } catch (error) {
    if(error.code === "ENOENT") {
      return false;
    } else {
      throw error;
    }
  }
}

function readFileSync(path, encoding) {
  return fs.readFileSync(path).toString(encoding);
}

function writeFileSync(path, data) {
  const stream = fs.openSync(path, "w");
  fs.writeSync(stream, data, 0, "utf8");
  fs.closeSync(stream);
}

export {
  existsSync,
  readFileSync,
  writeFileSync
};
```

```javascript
// elsewhere in your code where you need above file operations

// import { existsSync, writeFileSync, readFileSync } from "fs";
import { existsSync, writeFileSync, readFileSync } from "./fsExt.js";

...
```

p.s. above code doesn't fully wrap fs functions, it was enough in my case.

**Comment by tschaub** [comment]
@lg250137 - I added a test based on your example in https://github.com/tschaub/mock-fs/pull/392/commits/9c87858c9f4b5f9f66e36fde4e6f93b926aef345

This is passing now - likely fixed by #387.

Feel free to reopen if a more specific test cases fails.
